### ISSUE TITLE
C# project not noticing classes created within the same project

### ISSUE THREAD

**Issue by devsammi3** [premake/premake-core#1445]
I have a C# project defined this way:

```
project "CSharpStuff"
    location "CSharpStuff"
    language "C#"
    kind "SharedLib"
    
    dotnetframework "4.6.1"

    targetdir ("bin/" .. outputdir .. "/%{prj.name}")
    objdir ("bin/int/" .. outputdir .. "/%{prj.name}")

    files 
    {
        "%{prj.name}/**.cs",
    }

    filter "system:windows"
        systemversion "latest"

    filter "configurations:Debug"
        symbols "on"
        runtime "Debug"

    filter "configurations:Release"
        optimize "on"
        runtime "Release"

    filter "configurations:Dist"
        optimize "on"
        runtime "Release"
```

But files within that project are not noticing classes created in other files. This should have to do with an assembly reference I suppose but I don't know much about that and how to set that up using premake (I come from a C++ background). How would I go setting it up that I can use the same classes across different files?

I'm using premake5 btw

**Comment by samsinsane** [comment]
What you have looks pretty normal to me. Does Visual Studio offer any suggestions when you hover over the missing classes? In my experience, it's usually a missing namespace. Is it possible that you're missing other files for this project? Is this a brand new project, or are you trying to convert an existing project into Premake?

**Comment by starkos** [comment]
Are you able to create a minimal sample project that reproduces the problem? That would help a lot; I don't see anything immediately obvious in your example.

**Comment by devsammi3** [comment]
Yes, it's a simple project I'm creating to bind mono to C++ but, I'm not really focussing on it recognising the mono libraries right now, just trying to get a basic C# project started.

I have two files in the folder CSharpStuff; GameAssembly.cs and test.cs.

GameAssembly.cs looks like this:
```
using System;
using System.Timers;

namespace NURUEngine
{
    public class GameAssembly
    {
        static void Main()
        {
            Console.WriteLine("Hello, World!");
        }
    }
    
    public class Behaviour
    {
        [SerializeField]
        private uint myInt = 0;

        [SerializeField]
        private bool myBool = false;

        private Timer myTimer;

        Behaviour()
        {
            myInt = 3;
            myBool = true;
            myTimer = new Timer(1000);
            myTimer.Elapsed += OnTimedEvent;
            myTimer.AutoReset = true;
            myTimer.Enabled = true;
        }

        private static void OnTimedEvent(Object source, ElapsedEventArgs e)
        {
            Console.WriteLine("The Elapsed event was raised at {0:HH:mm:ss.fff}", e.SignalTime);
        }
    }
}
```

and test.cs like this:
```
using System;

namespace NURUEngine
{
    public class TestBehaviour : Behaviour
    {

    }
}
```
It doesn't even recognise Behaviour as a class in the test.cs file when I open it with visual studio

**Comment by samsinsane** [comment]
Visual Studio has an insane amount of diagnostics for C# projects compared to C++, in most cases you just need to hover over the broken code and it will offer up the solution. In your case it says:
![image](https://user-images.githubusercontent.com/11887457/81476819-47a49580-9257-11ea-8570-346e4170efb1.png)
Which means that `Behaviour` isn't marked as `public` - you've marked the class as `public`, which I believe means that other namespaces can know about it. But you haven't marked the constructor as `public`, so you get the above warning, changing it from `Behaviour()` to `public Behaviour()` will fix the problem.

**Comment by devsammi3** [comment]
![badbehaviour](https://user-images.githubusercontent.com/62988281/81477024-db597000-921d-11ea-8f55-3882a4cfdfc0.PNG)

Weird, for me, behaviour doesn't even show up as being recognised as a class.

Perhaps because it's in a solution with a C++ project? I doubt it but I'm just throwing a wild guess. The full premake file is this:

```
workspace "TestMono"

    architecture "x64"
    startproject "TestMono"

    configurations 
    {
        "Debug",
        "Release",
        "Dist"
    }

    flags
	{
		"MultiProcessorCompile"
    }
    
outputdir = "%{cfg.buildcfg}-%{cfg.system}-%{cfg.architecture}"

IncludeDir = {}
IncludeDir["Mono"] = "C:/Program Files/Mono/include/mono-2.0"

project "CSharpStuff"
    location "CSharpStuff"
    language "C#"
    kind "SharedLib"
    
    dotnetframework "4.6.1"

    targetdir ("bin/" .. outputdir .. "/%{prj.name}")
    objdir ("bin/int/" .. outputdir .. "/%{prj.name}")

    files 
    {
        "%{prj.name}/**.cs",
    }

    filter "system:windows"
        systemversion "latest"

    filter "configurations:Debug"
        symbols "on"
        runtime "Debug"

    filter "configurations:Release"
        optimize "on"
        runtime "Release"

    filter "configurations:Dist"
        optimize "on"
        runtime "Release"

project "TestMono"

    location "TestMono"
    kind "ConsoleApp"
    language "C++"
    cppdialect "C++17"

    targetdir ("bin/" .. outputdir .. "/%{prj.name}")
    objdir ("bin/int/" .. outputdir .. "/%{prj.name}")

    files 
    {
        "TestMono/**.h",
        "TestMono/**.cpp",
    }

    defines 
    {
        "_CRT_SECURE_NO_WARNINGS",
    }

    includedirs 
    {
        "%{IncludeDir.Mono}",
    }

    links 
    {
        "C:/Program Files/Mono/lib/mono-2.0-sgen.lib"
    }

    filter "system:windows"
        systemversion "latest"

    filter "configurations:Debug"
        symbols "on"
        runtime "Debug"

    filter "configurations:Release"
        optimize "on"
        runtime "Release"

    filter "configurations:Dist"
        optimize "on"
        runtime "Release"
```

**Comment by samsinsane** [comment]
Interesting, I've never had issues with C# projects in the same solution as C++ projects - we have a very simple sample project for our SDK, we actually have a public copy of it [here](https://github.com/Euclideon/vaultsdksamples/tree/master/languages/csharp) for what it's worth. I originally set it up in an internal repo that is entirely C++ projects and I vaguely recall having a similar tooltip with the protection level comment.

It's possible it's just a configuration problem, from memory you can trigger a partial install of the C# tools through the VS2019 IDE, and if you want the full setup you need to use the proper VS installer instead. Do you have the `.NET desktop development` workload installed?
![image](https://user-images.githubusercontent.com/11887457/81477235-0a8dd280-925a-11ea-9d0d-b9a95e0dec27.png)

**Comment by devsammi3** [comment]
Sorry, for the bother. I changed dotnetversion to 4.6 and it worked. Either .NET 4.6.1 doesn't exist or it isn't installed on my system for some reason. Everything now works. Final thing, if you're ever worked with mono and premake, anything I should look out for?

**Comment by samsinsane** [comment]
Very strange, I would say it's probably not installed since I copied the Premake script you provided at the top including using .NET 4.6.1. Since we've gotten to the bottom of the issue I'll close this off, feel free to open a new issue if you have any other issues or questions.

> Final thing, if you're ever worked with mono and premake, anything I should look out for?

Sorry, I haven't used Mono before - from my understanding the Mono team used (and might still be still be using) Premake to build Mono, so I can't imagine there being any major problems with using the two together. The only thing I can think of is the reference/assembly list in Visual Studio can be specified via `links` in Premake. For example, the sample I linked above uses `System.Drawing` and we linked that in `project.lua` with `links { "System.Drawing" }`, so you'll likely need to do something similar for the Mono assemblies.
